File: server/src/realtime/handleUplink.ts
// 网关上行帧的共用业务处理:被 HTTP 端点(/internal/gateway/uplink)与 gRPC 流服务(edgeGrpc)
// 共同复用——两条入口只差传输层,业务/落库/发号/幂等/扇出逻辑唯一一份(26-9-16 演进方案)。
// 身份(userId/deviceId)由网关注入(HTTP 头 / gRPC 帧头),不信任帧内自报的 senderId。
import type { Prisma } from '../generated/prisma/index.js';
import { sendMessageInput, readReportInput } from '../contracts/message.js';
import { buildDownlink } from '../contracts/ws.js';
import {
  persistMessage,
  getConversationMembers,
  markMentions,
} from '../services/message.js';
import { isConversationMember, advanceLastRead } from '../services/read.js';
import { filterOnline } from './presence.js';
import { handleCallEvent } from './callRelay.js';
import { redis } from '../database/redis.js';

const DOWNLINK_CHANNEL = 'gw:downlink';

// 把一条下行帧 publish 给指定用户(网关订阅 gw:downlink 后投给该用户在其副本的全部连接)。
export const publishDownlink = (
  userId: number,
  type: string,
  data: unknown,
  opts?: { exceptDeviceId?: string },
): Promise<number> =>
  redis.publish(DOWNLINK_CHANNEL, JSON.stringify(buildDownlink(userId, type, data, opts)));

// 从客户端上报的 mentions 里只保留「确实是本会话成员」的 id,杜绝跨会话伪造 @(与 socket.ts 同源约束)。
const parseMentionIds = (raw: unknown, participants: bigint[]): bigint[] => {
  if (!Array.isArray(raw)) return [];
  const memberSet = new Set(participants.map((p) => p.toString()));
  const out: bigint[] = [];
  for (const v of raw) {
    const s = String(v);
    if (/^\d+$/.test(s) && memberSet.has(s)) out.push(BigInt(s));
  }
  return out;
};

export interface UplinkContext {
  userId: number;
  deviceId: string;
}

// 处理结果对齐 HTTP 语义:204=不回投(仅确认);200=回投 body;其余=错误。
export interface UplinkResult {
  status: number;
  body?: unknown;
}

/** 处理一条网关上行帧(message.send / read.report / call:*)。 */
export async function handleUplink(frame: unknown, ctx: UplinkContext): Promise<UplinkResult> {
  const env = frame as { type?: string; data?: unknown } | undefined;
  const frameType = env?.type;
  if (!frameType) {
    return { status: 400, body: { type: 'message.error', message: '缺少帧类型 type' } };
  }

  // ===== 发消息:落库 + 读扩散扇出(网关侧唯一业务分支之外的既有权重路径) =====
  if (frameType === 'message.send') {
    // 兼容两种上行帧形态:
    //  ① {type:'message.send', data:{...消息载荷}} —— 推荐:信封 type 与消息载荷分离;
    //  ② 直接整帧即消息载荷(早期 PoC 形态)—— 兼容保留。
    const payload = (env.data ?? frame) as unknown;
    const parsed = sendMessageInput.safeParse(payload);
    if (!parsed.success) {
      return {
        status: 400,
        body: {
          type: 'message.error',
          message: '消息参数非法',
          data: { clientMsgId: (frame as { clientMsgId?: string })?.clientMsgId },
        },
      };
    }
    const data = parsed.data;

    try {
      const senderId = BigInt(ctx.userId);
      const participantIds = await getConversationMembers(data.conversationId, senderId);
      const { message, deduped } = await persistMessage({
        conversationId: data.conversationId,
        senderId,
        clientMsgId: data.clientMsgId,
        content: data.content,
        type: data.type,
        mentions: data.mentions as Prisma.InputJsonValue,
        extra: data.extra as Prisma.InputJsonValue,
        fileInfo: data.fileInfo as Prisma.InputJsonValue,
        participantIds,
      });

      // 去重命中不重复扇出(对方已收过首次广播),仅回 ack 让发送方收敛本地状态。
      if (!deduped) {
        const isGroup = data.conversationId.startsWith('group_');
        const targets = isGroup
          ? await filterOnline(participantIds)
          : new Set(participantIds.map(Number));
        await Promise.all(
          [...targets].map((uid) => publishDownlink(uid, 'receiveMessage', message))
        );

        const mentioned = parseMentionIds(data.mentions, participantIds);
        if (mentioned.length) {
          await markMentions(data.conversationId, message.seq, mentioned);
          const onlineMentioned = await filterOnline(mentioned);
          await Promise.all(
            [...onlineMentioned].map((uid) =>
              publishDownlink(uid, 'mention', {
                conversationId: data.conversationId,
                seq: message.seq,
                serverMsgId: message.id,
              })
            )
          );
        }
      }

      return {
        status: 200,
        body: {
          type: 'message.ack',
          data: {
            clientMsgId: data.clientMsgId,
            seq: message.seq,
            serverMsgId: message.id,
          },
        },
      };
    } catch (err) {
      console.error('网关上行处理失败:', err);
      return {
        status: 500,
        body: { type: 'message.error', data: { message: '消息发送失败', clientMsgId: data.clientMsgId } },
      };
    }
  }

  // ===== 已读上报:单调推进用户级 lastReadSeq,再推 read.sync 给同用户其它设备 =====
  if (frameType === 'read.report') {
    const parsed = readReportInput.safeParse(env.data);
    if (!parsed.success) {
      return { status: 400, body: { type: 'message.error', message: '已读上报参数非法' } };
    }
    const { conversationId, uptoSeq } = parsed.data;
    const userIdBig = BigInt(ctx.userId);
    try {
      if (!(await isConversationMember(userIdBig, conversationId))) {
        return { status: 403, body: { type: 'message.error', message: '无权操作该会话' } };
      }
      const { advanced } = await advanceLastRead(userIdBig, conversationId, uptoSeq);
      // 单调未推进(乱序旧值)时无需扰动其它端。
      if (advanced) {
        await publishDownlink(ctx.userId, 'read.sync', { conversationId, uptoSeq }, {
          exceptDeviceId: ctx.deviceId,
        });
      }
      return { status: 204 };
    } catch (err) {
      console.error('已读上报处理失败:', err);
      return { status: 500, body: { type: 'message.error', message: '已读上报失败' } };
    }
  }

  // ===== 通话信令:忙线裁决/属主路由/grace 重连,状态机在 services/callSession =====
  if (frameType.startsWith('call:')) {
    try {
      return await handleCallEvent(
        frameType,
        ctx.userId,
        ctx.deviceId,
        (env.data ?? {}) as Record<string, unknown>,
      );
    } catch (err) {
      console.error('网关信令处理失败:', err);
      return { status: 500, body: { type: 'call:error', message: '信令处理失败' } };
    }
  }

  return { status: 400, body: { type: 'message.error', message: `不支持的上行类型: ${frameType}` } };
}

File: biz/internal/realtime/handleUplink.go
package realtime

import (
	"context"
	"encoding/json"
	"time"

	"github.com/our-chat/biz/internal/metrics"
	"github.com/our-chat/biz/internal/service"
)

// handleUplink.go:网关上行帧共用业务处理,语义对齐 server/src/realtime/handleUplink.ts。
// HTTP 端点(/internal/gateway/uplink)与 gRPC 流服务共用同一份实现。

// UplinkContext 网关注入的身份(不信任帧内自报)。
type UplinkContext struct {
	UserID   int64
	DeviceID string
}

// UplinkResult 处理结果对齐 HTTP 语义:NoContent=不回投(仅确认);其余=回投 body。
type UplinkResult struct {
	Status int
	Body   any
}

// wsEnvelope 客户端信封 {type,data}。
type wsEnvelope struct {
	Type string          `json:"type"`
	Data json.RawMessage `json:"data"`
}

// HandleUplink 处理一条网关上行帧(message.send / read.report / call:*)。见 handleUplink.ts:52-181。
func HandleUplink(ctx context.Context, raw []byte, uctx UplinkContext) UplinkResult {
	var env wsEnvelope
	if err := json.Unmarshal(raw, &env); err != nil || env.Type == "" {
		return UplinkResult{Status: 400, Body: map[string]any{"type": "message.error", "message": "缺少帧类型 type"}}
	}

	switch {
	case env.Type == "message.send":
		return handleMessageSend(ctx, env.Data, raw, uctx)
	case env.Type == "read.report":
		return handleReadReport(ctx, env.Data, uctx)
	case len(env.Type) > 5 && env.Type[:5] == "call:":
		return HandleCallEvent(ctx, env.Type, uctx, env.Data)
	default:
		return UplinkResult{Status: 400, Body: map[string]any{"type": "message.error", "message": "不支持的上行类型: " + env.Type}}
	}
}

// ---- message.send(handleUplink.ts:60-137) ----

// sendMessageInput 上行发消息入参(contracts/message.ts 的 sendMessageInput 等价)。
type sendMessageInput struct {
	ClientMsgID    string          `json:"clientMsgId"`
	ConversationID string          `json:"conversationId"`
	SenderID       json.RawMessage `json:"senderId"` // 仅前端回显,忽略
	Content        string          `json:"content"`
	Type           string          `json:"type"`
	Mentions       json.RawMessage `json:"mentions"`
	Extra          json.RawMessage `json:"extra"`
	FileInfo       json.RawMessage `json:"fileInfo"`
}

// parseSendMessage 校验入参(默认值对齐 zod default 语义)。
func parseSendMessage(payload json.RawMessage) (sendMessageInput, bool) {
	var in sendMessageInput
	if err := json.Unmarshal(payload, &in); err != nil {
		return in, false
	}
	if in.ClientMsgID == "" || len(in.ClientMsgID) > 64 {
		return in, false
	}
	if in.ConversationID == "" || len(in.ConversationID) > 100 {
		return in, false
	}
	if in.Type == "" {
		in.Type = "text"
	}
	if len(in.Type) > 32 {
		return in, false
	}
	return in, true
}

func handleMessageSend(ctx context.Context, data json.RawMessage, raw []byte, uctx UplinkContext) UplinkResult {
	// 兼容两种上行帧形态:①{type,data} 推荐;②直接整帧即消息载荷(早期 PoC,handleUplink.ts:61-64)
	payload := data
	if len(payload) == 0 {
		payload = raw
	}
	in, ok := parseSendMessage(payload)
	if !ok {
		// 错误回投带 clientMsgId(handleUplink.ts:66-75)
		var cmid struct {
			ClientMsgID string `json:"clientMsgId"`
		}
		_ = json.Unmarshal(raw, &cmid)
		_ = json.Unmarshal(data, &cmid)
		return UplinkResult{Status: 400, Body: map[string]any{
			"type": "message.error", "message": "消息参数非法",
			"data": map[string]any{"clientMsgId": cmid.ClientMsgID},
		}}
	}

	metrics.MessageInTotal.Inc()
	start := time.Now()
	defer func() {
		metrics.MessageDuration.Observe(time.Since(start).Seconds())
	}()

	senderID := uctx.UserID
	participants, err := service.GetConversationMembers(ctx, in.ConversationID, senderID)
	if err != nil {
		metrics.MessageOutTotal.WithLabelValues("error").Inc()
		return UplinkResult{Status: 500, Body: map[string]any{
			"type": "message.error", "data": map[string]any{"message": "消息发送失败", "clientMsgId": in.ClientMsgID},
		}}
	}

	res, err := service.PersistMessage(ctx, service.PersistMessageInput{
		ConversationID: in.ConversationID,
		SenderID:       senderID,
		ClientMsgID:    in.ClientMsgID,
		Content:        in.Content,
		Type:           in.Type,
		Mentions:       in.Mentions,
		Extra:          in.Extra,
		FileInfo:       in.FileInfo,
		ParticipantIDs: participants,
	})
	if err != nil {
		metrics.MessageOutTotal.WithLabelValues("error").Inc()
		return UplinkResult{Status: 500, Body: map[string]any{
			"type": "message.error", "data": map[string]any{"message": "消息发送失败", "clientMsgId": in.ClientMsgID},
		}}
	}
	metrics.MessageOutTotal.WithLabelValues("ok").Inc()

	// 去重命中不重复扇出(handleUplink.ts:94-117)
	if !res.Deduped {
		isGroup := len(in.ConversationID) > 0 && in.ConversationID[0] == 'g'
		targets := map[int64]bool{}
		if isGroup {
			online, ferr := service.FilterOnline(ctx, participants)
			if ferr != nil {
				metrics.MessageOutTotal.WithLabelValues("error").Inc()
				return UplinkResult{Status: 500, Body: map[string]any{
					"type": "message.error", "data": map[string]any{"message": "消息发送失败", "clientMsgId": in.ClientMsgID},
				}}
			}
			targets = online
		} else {
			for _, p := range participants {
				targets[p] = true
			}
		}
		targetIDs := make([]int64, 0, len(targets))
		for uid := range targets {
			targetIDs = append(targetIDs, uid)
		}
		service.FanoutDownlink(ctx, targetIDs, "receiveMessage", res.Message)

		mentioned := service.ParseMentionIDs(in.Mentions, participants)
		if len(mentioned) > 0 {
			if err := service.MarkMentions(ctx, in.ConversationID, res.Message.Seq, mentioned); err != nil {
				logWarn("markMentions 失败", err)
			}
			onlineMentioned, _ := service.FilterOnline(ctx, mentioned)
			mentionIDs := make([]int64, 0, len(onlineMentioned))
			for uid := range onlineMentioned {
				mentionIDs = append(mentionIDs, uid)
			}
			service.FanoutDownlink(ctx, mentionIDs, "mention", map[string]any{
				"conversationId": in.ConversationID,
				"seq":            res.Message.Seq,
				"serverMsgId":    res.Message.ID,
			})
		}
	}

	return UplinkResult{Status: 200, Body: map[string]any{
		"type": "message.ack",
		"data": map[string]any{
			"clientMsgId": in.ClientMsgID,
			"seq":         res.Message.Seq,
			"serverMsgId": res.Message.ID,
		},
	}}
}

// ---- read.report(handleUplink.ts:140-163) ----

func handleReadReport(ctx context.Context, data json.RawMessage, uctx UplinkContext) UplinkResult {
	var in struct {
		ConversationID string `json:"conversationId"`
		UptoSeq        any    `json:"uptoSeq"`
	}
	if err := json.Unmarshal(data, &in); err != nil {
		return UplinkResult{Status: 400, Body: map[string]any{"type": "message.error", "message": "已读上报参数非法"}}
	}
	if in.ConversationID == "" || len(in.ConversationID) > 100 {
		return UplinkResult{Status: 400, Body: map[string]any{"type": "message.error", "message": "已读上报参数非法"}}
	}
	upto, ok := parseNonnegAny(in.UptoSeq)
	if !ok {
		return UplinkResult{Status: 400, Body: map[string]any{"type": "message.error", "message": "已读上报参数非法"}}
	}

	member, err := service.IsConversationMember(ctx, uctx.UserID, in.ConversationID)
	if err != nil {
		return UplinkResult{Status: 500, Body: map[string]any{"type": "message.error", "message": "已读上报失败"}}
	}
	if !member {
		return UplinkResult{Status: 403, Body: map[string]any{"type": "message.error", "message": "无权操作该会话"}}
	}

	advanced, err := service.AdvanceLastRead(ctx, uctx.UserID, in.ConversationID, upto)
	if err != nil {
		return UplinkResult{Status: 500, Body: map[string]any{"type": "message.error", "message": "已读上报失败"}}
	}
	// 单调未推进时无需扰动其它端(handleUplink.ts:152-157)
	if advanced {
		if err := service.PublishDownlink(ctx, uctx.UserID, "read.sync",
			map[string]any{"conversationId": in.ConversationID, "uptoSeq": upto},
			&service.DownlinkOpts{ExceptDeviceID: uctx.DeviceID}); err != nil {
			logDownlinkWarn(err)
		}
	}
	return UplinkResult{Status: 204}
}
